processor/tailsamplingprocessor/internal/sampling/percentage_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sampling
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+)
+
+func TestNewPercentageFilter_errorHandling(t *testing.T) {
+	_, err := NewPercentageFilter(zap.NewNop(), -1)
+	assert.EqualError(t, err, "expected a percentage between 0 and 1")
+
+	_, err = NewPercentageFilter(zap.NewNop(), 1.5)
+	assert.EqualError(t, err, "expected a percentage between 0 and 1")
+}
+
+func TestPercentageSampling(t *testing.T) {
+	var empty = map[string]pdata.AttributeValue{}
+
+	cases := []float32{0.01, 0.1, 0.125, 0.33, 0.5, 0.66}
+
+	for _, percentage := range cases {
+		t.Run(fmt.Sprintf("sample %.2f", percentage), func(t *testing.T) {
+			trace := newTraceStringAttrs(empty, "example", "value")
+			traceID := pdata.NewTraceID([16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16})
+
+			percentageFilter, err := NewPercentageFilter(zap.NewNop(), percentage)
+			assert.NoError(t, err)
+
+			traceCount := 2000
+			sampled := 0
+
+			for i := 0; i < traceCount; i++ {
+				decision, err := percentageFilter.Evaluate(traceID, trace)
+				assert.NoError(t, err)
+
+				if decision == Sampled {
+					sampled++
+				}
+			}
+
+			assert.InDelta(t, percentage*float32(traceCount), sampled, 0.001, "Amount of sampled traces")
+		})
+	}
+}
+
+func TestPercentageSampling_ignoreAlreadySampledTraces(t *testing.T) {
+	var empty = map[string]pdata.AttributeValue{}
+
+	trace := newTraceStringAttrs(empty, "example", "value")
+	traceID := pdata.NewTraceID([16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16})
+
+	var percentage float32 = 0.33
+
+	percentageFilter, err := NewPercentageFilter(zap.NewNop(), percentage)
+	assert.NoError(t, err)
+
+	traceCount := 100
+	sampled := 0
+
+	for i := 0; i < traceCount; i++ {
+		trace.Decisions = []Decision{NotSampled, NotSampled}
+		decision, err := percentageFilter.Evaluate(traceID, trace)
+		assert.NoError(t, err)
+
+		if decision == Sampled {
+			sampled++
+		}
+
+		// trace has been sampled, should be ignored
+		trace.Decisions = []Decision{NotSampled, Sampled}
+		decision, err = percentageFilter.Evaluate(traceID, trace)
+		assert.NoError(t, err)
+		assert.Equal(t, decision, NotSampled)
```

Comment:
This test here is a bit confusing, running the same trace against the same filter twice. It would be easier to understand the expected behavior if you created a table-test with scenarios, like these:

1. 100 spans, 90 errors, 10% sampling => 90 + 1 traces
2. 100 spans, 10 errors, 10% sampling => 10 + 9 traces